db/db_compaction_test.cc

```diff
@@ -7602,6 +7604,72 @@ class DBCompactionTestL0FilesMisorderCorruption : public DBCompactionTest {
   std::shared_ptr<test::SleepingBackgroundTask> sleeping_task_;
 };
 
+TEST_F(DBCompactionTest, CompactFilesSupportKeyPlacementRangeConflict) {
+  Options options;
+  options.create_if_missing = true;
+  options.compaction_style = kCompactionStyleUniversal;
+  options.num_levels = 3;
+  DestroyAndReopen(options);
+
+  // To create LSM of below shape:
+  // L0: [k2]
+  // L1: [k3],[k4]
+  // L2: [k1,  k5]
+  ASSERT_OK(Put("k1", "v"));
+  ASSERT_OK(Put("k5", "v"));
+  ASSERT_OK(Flush());
+  CompactRangeOptions cro;
+  cro.bottommost_level_compaction = BottommostLevelCompaction::kForceOptimized;
+  ASSERT_OK(db_->CompactRange(cro, nullptr, nullptr));
+  ASSERT_EQ("0,0,1", FilesPerLevel());
+
+  ASSERT_OK(Put("k3", "v"));
+  ASSERT_OK(Flush());
+  ASSERT_OK(Put("k4", "v"));
+  ASSERT_OK(Flush());
+  ASSERT_OK(experimental::PromoteL0(db_, db_->DefaultColumnFamily(), 1));
+  ASSERT_EQ("0,2,1", FilesPerLevel());
+
+  ASSERT_OK(Put("k2", "v"));
+  ASSERT_OK(Flush());
+  ASSERT_EQ("1,2,1", FilesPerLevel());
+
+  Close();
+
+  // To force below two CompactFiles() in order to coerce range conflict on L1
+  // upon (2)
+  // (1): Compact [k2] at L0 and [k3] at L1 and output to L1 and L0
+  // (2): Compact [k4] at L1 and [k1, k5] at L2 and output to L1 and L2
```

Comment:
In this case, it does not seem like an actual conflict: universal with preclude feature should only output to the penultimate level the keys that are within the key range of input files from penultimate level: https://github.com/facebook/rocksdb/blob/9bddac0dcfb6ea7fc55e64f41a2670cbc23c4a2d/db/compaction/compaction_iterator.cc#L1305-L1306. So only k4 can be potentially output to L1.

However, I think this change is fine since this seems to be an existing behavior.

Reply:
> In this case, it does not seem like an actual conflict: universal with preclude feature should only output to the penultimate level the keys that are within the key range of input files from penultimate level:

Oh interesting .... the conflict check happened before the compaction iterator so it crashed prematurely (or now aborted prematurely). Will leave a TODO